sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -1001,6 +1001,16 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
 }
 
 object CodeGenerator extends Logging {
+
+  // This is the value of `HugeMethodLimit` in JVM settings. Since we can't get this value on
+  // runtime and this is fixed in released JVMs, we hard-code this values here:
+  //
+  // $ java -XX:+UnlockDiagnosticVMOptions -XX:HugeMethodLimit=99999
+  // Error: VM option 'HugeMethodLimit' is develop and is available only in debug version of VM.
+  // Error: Could not create the Java Virtual Machine.
+  // Error: A fatal exception has occurred. Program will exit.
+  private val hugeMethodLimit = 8000
```

Comment:
Actually this threshold is only meaningful on HotSpot VM and some HotSpot-derived JVMs. Other JVMs, for instance IBM J9 doesn't use the same threshold. It'd be a bit too strict and biased to make this a non-configurable behavior for all JVMs.

I'd suggest that if we are to do this, at least centralize the JVM detection logic somewhere (e.g. unify with the JVM detection logic in `org.apache.spark.util.SizeEstimator`) and only set this kind of threshold based on the detected JVM. That way we're much less likely to regress on other JVMs, and/or even on future versions of the HotSpot VM where it'll get a new JIT compiler (Graal) with new JIT compilation heuristics that could be different from the current version.

Reply:
Thanks for your comment! yea, I agree. I' rethink this pr based on your suggestion.

Reply:
BTW, IBM J9 has the same threshold for too-long functions?

Reply:
Let's ask @kiszk for that ^_^
With [OpenJ9](http://openj9.mybluemix.net/) coming out soon, we might be able to find out details on this kind of issue ourselves. But until then it's easier to just ask IBM folks for a reliable answer ;-)

Reply:
oh, you know @kiszk, haha.

Reply:
While IBM JDK has [the similar option](http://www-01.ibm.com/support/docview.wss?uid=swg1IV78208) `-Xjit:acceptHugeMethods`, IBM JDK unfortunately does not disclose its threshold now.

My suggestion is to make this threshold configurable by using `SQLConf`.

Reply:
I see. I feel detecting specific JVM implementations might go too far for this purpose. yea, one of options is to add an internal option for this. WDYT? @rednaxelafx

Reply:
Sure, I'm fine with having a `SQLConf` conf flag for the threshold for now, with the option to move to a centralized JVM detection implementation later where we can still have a `SQLConf` flag to override the heuristic based on detection.
Setting such a threshold to 64KB will effectively turn the restriction off (i.e. same behavior as `-XX:-DontCompileHugeMethods`), so having a simple int conf should be good enough.

Reply:
yea, ok! Thanks!